### PR TITLE
tests/lib/tools/tests.invariant: add invariant for detecting broken snaps

### PR DESCRIPTION
**Pull request by anonymouse64**
Tests should generally not leave broken snaps lying around, if there are broken
snaps left, then they should be cleaned up by the test itself to avoid failing
this invariant.

I expect the initial opening of this to reveal some lurking dragons so opening as
a draft to see what happens and what tests need to be cleaned up.

### CODE REVIEW

File: tests/lib/tools/tests.invariant
Comment (MiguelPires) [inline tests/lib/tools/tests.invariant:130]:
```suggestion
		for broken in $(snap list --all | tail -n +2 | awk '{print $(NF) " " $1}' | grep 'broken ' | awk '{print $(NF)}' | sort | uniq); do
```

File: tests/lib/tools/tests.invariant
Comment (MiguelPires) [inline tests/lib/tools/tests.invariant:132]:
```suggestion
			echo "snap $broken is broken"
```

File: tests/lib/tools/tests.invariant
Comment (MiguelPires) [inline tests/lib/tools/tests.invariant:134]: Why write it to a file? Is this something we want to persist?

File: tests/lib/tools/tests.invariant
Comment (anonymouse64) [reply tests/lib/tools/tests.invariant:132]: ha I actually meant this to be a short-hand for "broken snap" but I can see how that's confusing, I will change it

File: tests/lib/tools/tests.invariant
Comment (anonymouse64) [reply tests/lib/tools/tests.invariant:134]: no, but that's just how all the other invariants are implemented and I followed the pattern

Comment (MiguelPires) [reply tests/lib/tools/tests.invariant:134]: I see, thanks for clarifying

**Comment by anonymouse64** [comment]
Surprisingly, no tests failed, but I can see from another test that failed that it supposedly is being run:
```
...
tests.invariant: broken-snaps ok
```

I'm going to push a purposefully broken spread test to ensure that the broken-snaps invariant actually picks it up, since maybe I have the bash wrong and it's not actually catching the broken snap.

File: tests/lib/tools/tests.invariant
Comment (bboozzoo) [inline tests/lib/tools/tests.invariant]: Hmm did you mean something like:
```
# fist column is the snap name, revision is 3rd from the end, we cannot use $3 as
# broken snaps have empty version column
snap list --all | awk '/,?broken,?/ { print $1, $(NF-3) }' | while read -r name rev; do 
    echo "snap $name ($rev) is broken"
done
```
With the dump from debug log we get:
```
snap core18 (1888) is broken
snap pi (74) is broken
snap snapd (8791) is broken
```

On a side note, perhaps we should really print some placeholder for a version if the snap is broken. We use `-` for the tracking and publisher columns when the snap is unasserted. cc @pedronis

Comment (bboozzoo) [reply tests/lib/tools/tests.invariant]: Added a placeholder in https://github.com/snapcore/snapd/pull/11166

Comment (anonymouse64) [reply tests/lib/tools/tests.invariant]: It was indeed my intention to have the revision number also printed but since I couldn't figure out a nice way to get that due to the missing version in the second column I resorted to this hackery but your hackery is more useful as it also has the revision number :-) Thanks for the suggestion

File: tests/lib/tools/tests.invariant
Comment (stolowski) [inline tests/lib/tools/tests.invariant:130]: This requires a running snapd instance, I wonder if it wouldn't be desirable to have a custom check that doesn't require snapd (this most likely means checking that snap.yaml is readable for all `/snap/*/` snaps?)?

Comment (anonymouse64) [reply tests/lib/tools/tests.invariant:130]: Well slightly later on during the restore we will remove all snaps by calling `snap remove` (that actually was what triggered me to want to write this since that snippet was failing because of broken snaps), so I think it's a fair requirement for snapd to be running after a test completes

File: tests/lib/tools/tests.invariant
Comment (anonymouse64) [reply tests/lib/tools/tests.invariant]: Added this suggestion since the other PR has not yet landed

Comment (anonymouse64) [reply tests/lib/tools/tests.invariant]: Hmm, I added this suggestion but now the tests are not failing, I think there is something wrong with this somewhere... Feel free to push things to this branch to try other things, or we could just wait until that PR has landed to do the easy and simple implementation

File: tests/lib/tools/tests.invariant
Comment (bboozzoo) [inline tests/lib/tools/tests.invariant:135]: I ran this manually after having unmounted some snaps, the check seems to work:
```
google:ubuntu-20.04-64 .../tests/main/snap-info# tests.invariant check
tests.invariant: root-files-in-home ok
tests.invariant: crashed-snap-confine ok
tests.invariant: lxcfs-mounted ok
tests.invariant: stray-dbus-daemon ok
tests.invariant: leftover-defer-sh ok
tests.invariant: broken snaps
snap core (12383) is broken
snap test-snapd-devmode (1) is broken
snap test-snapd-tools (6) is broken
tests.invariant: broken-snaps not-ok
tests.invariant: system is corrupted
```

**Review by bboozzoo — APPROVED** [review]
Thanks

**Review by MiguelPires — APPROVED** [review]
LGTM

**Review by stolowski — APPROVED** [review]
Looks good and useful, thanks.

**Comment by bboozzoo** [comment]
@mvo5 @pedronis this is ready to land, can you merge it?

File: tests/lib/tools/tests.invariant
Comment (anonymouse64) [reply tests/lib/tools/tests.invariant]: followup in https://github.com/snapcore/snapd/pull/11204

**Pull request merged**
